### DISCUSSION TITLE
Revisiting the branching strategy

### DISCUSSION THREAD

**Discussion by rossabaker** [http4s/http4s#4106] [Ideas]
# Current branches

| Branch              | Merges from            | Target release | Cats Effect | Scala | Breaking?       |
|---------------------|---------------------|----------------|-------------|-------|-----------------|
| series/0.21         |                     | 0.21           | CE2         | 2     | No              |
| dotty               | series/0.21         | 0.21           | CE2         | 2,3   | private[http4s] |
| main                | series/0.21         | 1.0.0-M*       | CE2         | 2     | Yes             |
| cats-effect-3       | main                | 1.0.0          | CE3         | 2     | Yes             |
| cats-effect-3-dotty | dotty,cats-effect-3 | 1.0.x          | CE3         | 2,3   | private[http4s] |

Pros:
* People might get on Dotty milestones without also taking on Cats-Effect 3. It has been hell on us.
* Other major players (e.g., Cats Effect 2, FS2-2) support Dotty in production, and we could, too.

Cons:
* The new parsers on `dotty` can't correct model deficiencies
* Few people are working on `main`. Most of its motivators are on `dotty` or `cats-effect-3`.
* We have 65 and counting MiMa exceptions on Dotty. They're all `private[http4s]`, but, yikes!
* The first 0.21 release after dotty is merged may be unstable after parsers are rewritten
* There's interest in cats-uri, but we're putting people off as cats-parse support percolates through the 0.21 model
* Five branches is bonkers.

# Revised proposal

| Branch              | Merges from            | Target release | Cats Effect | Scala | Breaking? |
|---------------------|---------------------|----------------|-------------|-------|-----------|
| series/0.21         |                     | 0.21           | CE2         | 2     | No        |
| main                | series/0.21         | 1.0.0-M*       | CE2         | 2,3     | Yes       |
| cats-effect-3       | main                | 1.0.0          | CE3         | 2     | Yes       |

Pros:
* Long-running `dotty` is eliminated. `cats-parse` work moves to `main`, and Dotty support will be introduced as a pull request into a releasable branch.
* Parser work could directly include model cleanups instead of requiring multiple steps
* We could spin off cats-uri and enlist more volunteers sooner
* Flattens to two branches after CE3 port is done.

Cons:
* Destabilizes people already on 1.0 milestones in production.  (We warned you!)
* Goes back on our word. (Your money back!)
* Someone has a hellmerge from `dotty` to `main`. (Stand back, I have Emacs!) [#4108]
* Couples a stable CE3 release to cats-parse and maybe cats-uri

# Alternate proposal

| Branch              | Merges from          | Target release | Cats Effect | Scala | Breaking? |
|---------------------|---------------------|----------------|-------------|-------|-----------|
| series/0.21         |                     | 0.21           | CE2         | 2     | No        |
| series/0.22                | series/0.21                | 0.22          | CE3         | 2     | Yes       |
| main                | series/0.22         | 1.0       | CE3         | 2,3     | Yes       |

This recognizes that CE3 development is ahead of Dotty, and that our CE3 port is ahead of the Dotty port.  This race looked more even at the time of initial planning.

We would fork series/0.22 off main before merging #4108.
 
Pros:
* Stable CE3 release can be cut independently of ambitious projects like cats-parse and cats-uri
* We might get an 0.22-RC out while cats-effect is still in RC
* 1.0.0 can be held for Scala-3.0.0 so we can be sure it throws us no curveballs

Cons:
* Two breaking upgrades for users between today and http4s-1.0.
* Leaves three active branches until we're ready to EOL CE2.
* Confusing that 0.22.0 includes things first introduced in 1.0 milestones.
* Model fixes made before 0.22 need parboiled support

# New Year's Day proposal

| Branch              | Merges from            | Target release | Cats Effect | Scala | Breaking? |
|---------------------|---------------------|----------------|-------------|-------|-----------|
| series/0.21         |                     | 0.21           | CE2         | 2     | No        |
| series/0.22                | series/0.21         | 0.22       | CE2         | 2,3     | Yes       |
| main       | series/0.22                | 1.0.0          | CE3         | 2,3     | Yes       |

* Basically the Revised Proposal, but we release the dotty work on CE2.
* `dotty` is merged to `main` and becomes the new `series/0.22`. 
* `cats-effect-3` becomes the new `main` 
 
Pros:
* Long-running `dotty` is eliminated: cats-parse work is on a releasable branch.
* Users in the CE2 ecosystem can upgrade to Dotty without biting off CE3.
* Parser work could directly include model cleanups instead of requiring multiple steps.
* We could spin off cats-uri and enlist more volunteers sooner

Cons:
* Couples a stable CE3 release to cats-parse and maybe cats-uri
* Two breaking upgrades for users between today and http4s-1.0, though 0.22 is easy.
* Confusing that 0.22.0 includes things first introduced in 1.0 milestones.

**Comment by Daenyth** [comment]
I don't have a large stake since I have no ongoing work right now, but I'm in favor of whatever decreases the net maintenance burden on library maintainers and lets us iterate faster.

Users definitely shouldn't be using 1.0 milestones in production, since it's been so heavily in flux -- CE3 and fs2-3 still aren't even totally stable

**Reply by rossabaker** [reply]
1.0 milestones are currently on stable versions of cats-effect-2 and fs2-2.  But yeah, since we're planning on merging cats-effect-3 into that and calling that 1.0, I think your point still holds. People on that never got the binary compatibility promise of 0.21, which I think we've done (and continue to do) an outstanding job of.

**Comment by cquiroz** [comment]
It seems a good proposal. 👍

**Comment by rossabaker** [comment]
I updated the proposal to drop the `dotty` branch.  We can start moving Scala 2 macros over anytime, and Scala 3 macros can lie dormant until they're all done, at which point we add it back to CI.

**Comment by RaasAhsan** [comment]
What does the timeline for http4s-1.0.0 release look like with respect to Dotty/CE3? As per Daniel it sounds like CE3 is targeted to release before Dotty, so assuming there is no outstanding http4s work at that point, could you start cutting 1.x releases for CE3+Scala 2?

**Reply by rossabaker** [reply]
* We can start cutting milestones for CE3 + Scala 2 as soon as all the modules are ported.
* We could cut http4s-1.0.0 release anytime after cats-effect-3.0.0 and fs2-3.0.0, but...
  * This proposal makes cats-parse-1.0 another prerequisite of 1.0.0.  That's okay, because swapping out parser implementations is at best binary-breaking change to package privates, and risky to end users.  These need to bake for a bit.
  * If we spin off cats-uri, that's another prerequisite.  We need volunteers if that's going to happen.
  * 1.0 has been a proposed time to spin off integrations so we don't break when they break.  That addresses some other risks:
    * Circe, a major dependency, is itself 0.x.
    * Servlet 5 support?
* Scala 3 milestones can happen as soon as it compiles, before or after CE3 support is merged.
  * Scala 3 builds get no binary compatibility guarantee before scala-3.0.0

**Reply by rossabaker** [reply]
As I typed the above, I could already imagine high demand on gitter for a cats-effect-3 release, so I conceived Alternate Proposal above.

**Reply by rossabaker** [reply]
In this alternate proposal, I would suggest we could start cutting milestones of 0.22.x even before all the modules are ported.

**Reply by RaasAhsan** [reply]
Thanks, this is super comprehensive :) Alternate proposal looks compelling but I suppose it comes down to how confident we are that CE3 releases before Dotty, but also how much time is in between those two releases. We should try to get a sense of that, but Dotty is obviously the bigger unknown

**Comment by RaasAhsan** [comment]
In both the revised and alternate proposals, what are the risks in merging CE3 development into `main`? The obvious one to me is Dotty reaching stable before CE3, but anything in terms of maintenance burden?

**Reply by RaasAhsan** [reply]
I guess the obvious one is that you can't publish milestones yet since you can't run an http4s app on CE3! Is it worth considering doing sometime soon though? maybe once blaze-* and some of the other clients are done

**Reply by rossabaker** [reply]
Revised Proposal anticipates Dotty (including cats-parse) finishing before CE3 release.  If Dotty is released and cats-parse is done, we could tag main as 0.22 to deliver that, and let our CE3 finish at its own pace.  If CE3 is released first, we'd be the Shame of Typelevel being late to support it.

Alternate Proposal anticipates CE3 finishes before Dotty.  We could join the CE3 parade fairly quickly let Dotty finish at its own pace.  If Dotty is released first, we'd be held back by CE3.

In both proposals, the risk of being wrong is that we have to finish both major projects before delivering one.  If we're truly not sure, continuing to develop dotty and CE3 on separate branches is right.  We might probably just fold cats-effect into main now and make it a four-branch strategy: our 1.0 milestones are just whatever development isn't binary compatible.  But even at four branches, it's heavier than the other two proposals.

If CE3 before Dotty is highly likely, then Alternate Proposal feels right to me.

**Comment by RaasAhsan** [comment]
New Years Day proposal looks solid. Is it fair to say that:
1. You can start cutting 1.0 releases off of `main` as soon as CE3 drops, targeting Dotty milestones if necessary.
2. 1.x is basically blocked on CE3, so Dotty could drop and we don't have a 1.x release. I now realize this isn't necessarily a flaw.
3. You can start cutting 0.22 releases whenever the parser/model changes go through, regardless of whether Dotty has dropped or not.

**Reply by rossabaker** [reply]
> You can start cutting 1.0 releases off of main as soon as CE3 drops, targeting Dotty milestones if necessary.

Yes.

> 1.x is basically blocked on CE3, so Dotty could drop and we don't have a 1.x release. I now realize this isn't necessarily a flaw.

Yes, but I think CE3 will be done before Dotty anyway.

> You can start cutting 0.22 releases whenever the parser/model changes go through, regardless of whether Dotty has dropped or not.

Yes.  I think we should have a working build on a Dotty milestone before calling 0.22.0 _or_ 1.0.0 final, to have high confidence we can support Scala 3.0 without further breaking changes.

We also need to be mindful of cats-parse.  We're currently on 0.2 of that, and there are significant, unreleased, breaking changes.  If that's not 1.0 when we're otherwise ready, we need to consider shading it.

**Reply by hamnis** [reply]
I'm in favor of the New Years Day proposal.

**Reply by rossabaker** [reply]
Okay, let's move forward with that.
